# F-Test

## Assumptions

The F-Test (do not confuse with ANOVA) is used in this course as a tool for the decision which test is used for the comparison of two samples. It checks for heterogeneity of variances. The test result complements the consideration of boxplots.

The hypotheses for this test are called: 

$$H_{0}: \frac{\sigma_{A}}{\sigma_{B}} =  1$$

$$H_{1}: \frac{\sigma_{A}}{\sigma_{B}} \neq 1$$


<div class="alert alert-info"><b>Careful:</b>A significance in the F-test concludes a heterogeneity in variances. It is not possible to conclude a homogeneity from a non significant test result.</div>

Normal distribution of both samples is an important assumption for the F-test.

## Implementation

I am currently not aware of a pre-implemented function in e.g. Pandas, SciPy or NumPy. (Maybe I just didn't find it, it may exist.) Therefore, I here show you how to compute a two-sided F-Test in Python with a custom function that expects two lists of values for the two samples:


In [1]:
import numpy as np
import scipy
import scipy.stats

# Source: https://www.statology.org/f-test-python/
def f_test(x, y):
    x = np.array(x)
    y = np.array(y)
    f = np.var(x, ddof=1)/np.var(y, ddof=1) #calculate F test statistic 
    fdistribution = scipy.stats.f(len(x)-1,len(y)-1) # build an F-distribution object
    p = 2*min(fdistribution.cdf(f), 1-fdistribution.cdf(f))
    return f, p

You can call this function as 

    f_test(x_list, y_list)
    
## Example "Wisconsin Fast Plant"

### Experiment

"The "Wisconsin Fast Plant", *Brassica campestris*, has a very rapid growth cycle that makes it particularly well suited for the study of factors that affect plant growth. In one such study, seven plants were treated with the substance Ancymidol (ancy) and were compared to eight control plants that were given ordinary water. Heights of all of the plants were measured, in cm, after 14 days of growth" (see data below, Ahern, 1998, cited according to Samuels and Wittmer, 2003, p.228). Ancymidol is a growth suppressor used in agriculture as a herbicide. 

Are the variances homogeneous?

### Statistical Analysis  

In [2]:
import pandas as pd
brassica = pd.read_csv('data/brassica.csv', sep='\t')

In [3]:
f_test(brassica[brassica['group'] == "ancy"]['height'], brassica[brassica['group'] == "control"]['height'])

(0.9731551139130411, 0.9898470893751512)

### Interpretation

The F-Test statistic is 0.9731551139130411, and the p-value is 0.9898470893751512. The p-value is compared to an $\alpha$-error that has been set a priori. If the p-value is smaller than $\alpha$ then the alternative hypothesis will be accepted. 

The F-test checks for heterogeneity of variances. Although the homogeneity of variances is more interesting in this case, there is no test for homogeneity existing as far as I know. There are no general rules how to treat the output of a F-test when looking for homogeneity. I assume the variances to be more or less homogeneous if the p-value is rather big - including the interpretation of the boxplots. 

A p-value of 0.9898470893751512 implies, that there is no significant heterogeneity in variances (comparing with an $\alpha$ of 5\%) $\Longrightarrow$ homogeneity of variances.